hashing/src/chunk_with_proof.rs

```diff
@@ -0,0 +1,310 @@
+
+use casper_types::bytesrepr::{self, Bytes, FromBytes, ToBytes};
+
+use crate::{error::MerkleConstructionError, indexed_merkle_proof::IndexedMerkleProof, Digest};
+
+/// Represents a chunk of data with attached proof.
+#[derive(PartialEq, Debug, JsonSchema, Serialize, Deserialize)]
+#[schemars(with = "String", description = "Chunk of data with attached proof.")]
+#[serde(deny_unknown_fields)]
+pub struct ChunkWithProof {
+    proof: IndexedMerkleProof,
+    #[schemars(with = "String", description = "Hex-encoded bytes.")]
+    chunk: Bytes,
+}
+
+impl ToBytes for ChunkWithProof {
+    fn to_bytes(&self) -> Result<Vec<u8>, bytesrepr::Error> {
+        let mut result = bytesrepr::allocate_buffer(self)?;
+        result.append(&mut self.proof.to_bytes()?);
+        result.append(&mut self.chunk.to_bytes()?);
+        Ok(result)
+    }
+
+    fn serialized_length(&self) -> usize {
+        self.proof.serialized_length() + self.chunk.serialized_length()
+    }
+}
+
+impl FromBytes for ChunkWithProof {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), bytesrepr::Error> {
+        let (proof, remainder) = FromBytes::from_bytes(bytes)?;
+        let (chunk, remainder) = FromBytes::from_bytes(remainder)?;
+
+        Ok((ChunkWithProof { proof, chunk }, remainder))
+    }
+}
+
+impl ChunkWithProof {
+    #[cfg(test)]
+    /// 10 bytes for testing purposes.
+    pub const CHUNK_SIZE_BYTES: usize = 10;
+
+    #[cfg(not(test))]
+    /// 1 MiB
+    pub const CHUNK_SIZE_BYTES: usize = 1 << 20;
+
+    /// Constructs the [`ChunkWithProof`] that contains the chunk of data with the appropriate index
+    /// and the cryptographic proof.
+    ///
+    /// Empty data is always represented as single, empty chunk and not as zero chunks.
+    pub fn new(data: &[u8], index: u64) -> Result<Self, MerkleConstructionError> {
+        let (proof, chunk) = match data.is_empty() {
+            true if index > 0 => {
+                return Err(MerkleConstructionError::IndexOutOfBounds { count: 1, index });
+            }
+            false if index >= data.chunks(Self::CHUNK_SIZE_BYTES).len() as u64 => {
+                return Err(MerkleConstructionError::IndexOutOfBounds {
+                    count: data.chunks(Self::CHUNK_SIZE_BYTES).len() as u64,
+                    index,
+                })
+            }
+            true => (
+                IndexedMerkleProof::new([Digest::hash(&[])], index)?,
+                Bytes::new(),
+            ),
+            false => (
+                IndexedMerkleProof::new(
+                    data.chunks(Self::CHUNK_SIZE_BYTES).map(Digest::hash),
+                    index,
+                )?,
+                Bytes::from(
+                    data.chunks(Self::CHUNK_SIZE_BYTES)
+                        .nth(index as usize)
+                        .unwrap(), // We're safe to unwrap() because index is validated above
+                ),
+            ),
+        };
+
+        Ok(ChunkWithProof { proof, chunk })
+    }
```

Comment:
```suggestion
    pub fn new(data: &[u8], index: u64) -> Result<Self, MerkleConstructionError> {
        let chunk: Bytes;
        let proof: IndexedMerkleProof;
        if data.is_empty() {
            chunk = vec![].into();
            proof = IndexedMerkleProof::new([Digest::hash(&[])], index)?
        } else {
            chunk = data
                .chunks(Self::CHUNK_SIZE_BYTES)
                .nth(index as usize)
                .ok_or_else(|| MerkleConstructionError::IndexOutOfBounds {
                    count: data.chunks(Self::CHUNK_SIZE_BYTES).len() as u64,
                    index,
                })?
                .into();
            proof = IndexedMerkleProof::new(
                data.chunks(Self::CHUNK_SIZE_BYTES).map(Digest::hash),
                index,
            )?;
        };
        Ok(ChunkWithProof { proof, chunk })
    }
```

Reply:
I agree with this simplification: The `index` checks are duplicated; just use what `IndexedMerkleProof::new` returns.

Reply:
Now using `Digest::hash_merkle_tree()`.

Reply:
Right, I'd prefer documenting rather than duplicating the check. Sorry I didn't see that thread earlier! Sure, let's discuss tomorrow.

Reply:
@rafal-ch Probably best to leave resolving a comment thread to the individual who started the thread.

Reply:
> Now using `Digest::hash_merkle_tree()`.

You appear to be talking about https://github.com/casper-network/casper-node/pull/2119/commits/b4e3ccb8f5d4b39e3016fdd2d6403e9ffe9575d9 .  I think you mean `Digest::blake2b_hash` rather than `Digest::hash_merkle_tree()`, right?

Looking back at what you had, it looks a slightly more efficient than the golfed thing we have now:

```rust
     data[Self::CHUNK_SIZE_BYTES * (index as usize)
                    ..data
                        .len()
                        .min(Self::CHUNK_SIZE_BYTES * ((index as usize) + 1))]
                    .to_vec()
```

What was wrong with the old thing?

Reply:
Yes, indeed. I meant the `Digest::blake2b_hash`, sorry for confusion.

Regarding the code, in my opinion `.nth()` expresses the intent of the code much clearer. I'd suggest sticking to it unless proven significantly less efficient. I can consider writing benches for that, but this code is used only when constructing `ChunkWithProof` which - I believe - is not on the hot-path.

Let's discuss today as well, to avoid changing the code back and forth due to style preferences.

Reply:
As discussed, the code has been changed:
* No explicit index check for empty data (we rely on `IndexedMerkleProof::new()`)
* For non-empty case, `unwrap()` disappears and we use `ok_or_else()`

Changes are here: https://github.com/casper-network/casper-node/pull/2119/commits/0dc036e9424acbe9c3a735805947043579cdd0d2